Source/WebCore/platform/graphics/ComplexTextController.cpp

```diff
@@ -776,12 +776,17 @@ void ComplexTextController::adjustGlyphsAndAdvances()
             const auto& textAutoSpace =  m_font.textAutospace();
             float textAutoSpaceSpacing = 0;
             auto characterClass = TextSpacing::CharacterClass::Undefined;
-            if (!textAutoSpace.isNoAutospace())
+            // Since we are iterating through glyphs here we skip combining marks, since we just care about the grapheme cluster base for text-autospace.
+            if (!textAutoSpace.isNoAutospace() && !isCombiningMark(character)) {
                 characterClass = TextSpacing::characterClass(character);
-            if (textAutoSpace.shouldApplySpacing(previousCharacterClass, characterClass)) {
-                textAutoSpaceSpacing = complexTextRun.textAutospaceSize();
-                advance.expand(textAutoSpaceSpacing, 0);
+                if (textAutoSpace.shouldApplySpacing(previousCharacterClass, characterClass)) {
+                    textAutoSpaceSpacing = complexTextRun.textAutospaceSize();
+                    advance.expand(textAutoSpaceSpacing, 0);
+                }
+
+                previousCharacterClass = characterClass;
```

Comment:
Not sure it's a good idea to move this step... On the one hand, it packages up this logic nicely in one place; on the other it splits up the "let's capture the current state and move forward one step" part. Could go either way, so I'll leave that to you and your official reviewer. :)

Reply:
The reason for that is we are making a glyph walk here, not a code point/grapheme cluster walk. The previous character class is solely used for text-autospace computation. We should not update the previous character class when we passed through a combining mark because the adjacency should be computed between the glyph correspondent to the base character and the next glyph.